theme docs update (#5379)
vaadin/docs: articles/styling/themes/aura/color.adoc

Reviewer: Should this mention somewhere that Aura applies the blue accent color by default?

Author: Added.

---

= Aura Color

> Aura defines a set of color properties that define the look and feel of Vaadin components and allow for easy customization of the theme. They can also be applied to custom UI compositions instead of native CSS values.

Reviewer: Nit: "Native" sounds like there is something special going on, which it isn't. A custom CSS prop is "native" CSS. Maybe "static" or "constant" would be better in this context.

Author: Maybe this is clear enough?

```suggestion
Aura defines a set of color properties that define the look and feel of Vaadin components and allow for easy customization of the theme. They can also be used to style custom UI compositions.
```

Reviewer: Yep, that works

[source,css]
----
/* Use light/dark-suffixed properties for customization */
html {
  --aura-accent-color-light: #ffd06b;
  --aura-accent-color-dark: #1b0f38;
}

/* Use un-suffixed properties as values */
.some-ui-element {
  background-color: var(--aura-accent-color);
}
----

Any other color values you use won't automatically adapt to the color scheme, unless you provide separate values for each using the `light-dark()` CSS function:

[source,css]
----
.some-ui-element {
  background-color: light-dark(white, black);
}
----

[[light-dark-function]]
.Avoid light-dark() in Aura Customization
[CAUTION]
====
In Safari 17, the `light-dark()` function does not work with the way many Aura color properties are implemented. This limitation is most notable with the <<#saturated-palette,saturated palette colors>>, which need to use the same color in both light and dark color schemes if you want to support Safari 17.

You can use the `light-dark()` function to assign values to properties marked with [.badge.light-dark]#light-dark()# or xref:./#read-only-properties[Read-only,role=badge].

> ====

Reviewer: This whole admonition is confusing me:
- It says I can not use light-dark
- But only if I care about Safari 17
- But then it says I can use it
- To customize read-only properties, which some other part of the docs says I must not customize?

Can we find some more straight-forward advice that would apply to the average user?

Related to that I didn't understand the purpose of the light-dark badge, except that I'm probably not supposed to customize those, since rows in the table that have it use the same gray background that props with read-only have.

Author: I suppose we can avoid mentioning Safari 17, and just state that you can't use `light-dark()`, unless the property is explicitly marked to support that. Do you think what would work?

And yeah, such properties are by default computed from other properties, and the recommendation is that you shouldn't manually edit those. But you can, if you know what you’re doing. I suppose I struggle to define the target audience for this documentation. I wouldn't want to dumb it down too much, so that it no longer serves our own UX designers, who are very capable with CSS.

Reviewer: > I suppose we can avoid mentioning Safari 17, and just state that you can't use light-dark(), unless the property is explicitly marked to support that. Do you think what would work?

That would be a better start IMO. If we get a lot of feedback on this, we can revise to be more explicit.

Regarding the other point, I get the conflict between providing simple instructions and being thorough on all the capabilities of the system. I think what's missing at the moment is:
- A purpose as a "legend" that people can use to understand what it means when a property is marked as light-dark
- A similar paragraph as in `/styling/themes/aura#style-properties` that says that you should consider light-dark properties as read-only, but you can customize them if you know what you're doing. 

So maybe instead of having this admonition say "Avoid light-dark", align it more with the other admonition:
- Just title it as "Light-dark properties"
- Similar explanation that those are considered read-only, but you can customize them if you handle the details yourself
- Mention that using the light-dark CSS function is only supported in properties marked as light-dark

Author: Updated

== Background Color

.Surface Color
[TIP]
In addition to the style properties listed here, Aura defines many background colors with the <<#surface-color,surface color>>.

[.property-table]
`--aura-app-background` xref:#light-dark-function[light-dark(),role="badge light-dark"]::
The main background image, used for the `<html>` element and the App Layout content area. By default this is a computed gradient, based on the `--aura-background-color` property, which becomes more prominent with more saturated background colors. You can also set it to a solid color as well.

`--aura-background-color` xref:./#read-only-properties[Read-only,role=badge]::

> The primary background color that defines the overall tone of the application. This color is used to computed <<#text-color,text>>, <<#border-color,border>>, and <<#surface-color,surface>> colors. Automatically assigned the value of the `-light` or `-dark` suffixed property, depending on the active color scheme.

Reviewer: ```suggestion
The primary background color that defines the overall tone of the application. This color is used to compute <<#text-color,text>>, <<#border-color,border>>, and <<#surface-color,surface>> colors. Automatically assigned the value of the `-light` or `-dark` suffixed property, depending on the active color scheme.
```

== Text Color

Aura uses the `--vaadin`-prefixed <<../base#text-color,base style text properties>> by assigning its own computed, color-scheme-adapted values to them. Aura also defines a text color that is tinted with the <<accent-color>> as well as text colors for each of the <<#palette,palette colors>>.

[.property-table]
`--vaadin-text-color` xref:#light-dark-function[light-dark(),role="badge light-dark"]::
The main text color used in the UI. Defaults to the <<#neutral-color,neutral color>>.

`--vaadin-text-color-secondary` xref:#light-dark-function[light-dark(),role="badge light-dark"]::
A secondary text color, with at least 4.5:1 contrast ratio with the main background color. Computed from `--vaadin-text-color` by default.

`--vaadin-text-color-disabled` xref:#light-dark-function[light-dark(),role="badge light-dark"]::
Text color for disabled text. Computed from `--vaadin-text-color` by default.

`--aura-accent-text-color` xref:./#read-only-properties[Read-only,role=badge]::
A text color derived from the accent color, with at least 4.5:1 contrast ratio with the main background color.

`--aura-accent-text-color` xref:./#read-only-properties[Read-only,role=badge]::

> A text color derived from the accent color, with at least 4.5:1 contrast ratio with `--aura-background-color`.

Reviewer: This one is duplicated.

}
----

.Using the orange accent color in the light color scheme and the purple accent color in the dark color scheme
image::_images/accent-colors.png[]
====

=== Accent Color Class Names

The accent color is applied globally to the entire UI, but you can also scope it to individual elements.

Aura provides the following CSS utility classes, based on the <<#saturated-palette,saturated color palette>>, that you can apply to any part of the UI to give it a different accent color:

- `aura-accent-neutral`
- `aura-accent-red`
- `aura-accent-orange`
- `aura-accent-yellow`
- `aura-accent-green`
- `aura-accent-blue`
- `aura-accent-purple`

// .Use the purple palette color as the global accent color
// [example]
// ====
// [source,html]
// ----
// <html class="aura-accent-purple">
// ----
// ====

.Use the purple palette color as the accent color of a button
[example]
====
[.example]
--
[source,java]
----
<source-info group="Java"></source-info>
var button = new Button("Generate with AI");
button.addClassNames("aura-accent-purple");
----
[source,html]
----
<source-info group="HTML"></source-info>
<vaadin-button class="aura-accent-purple">Generate with AI</button>
----
--
====

> Furthermore, you can use the `-aura-accent-color` class name to reset the accent color back to global accent color, for example, if a parent container has applied another accent color.

Reviewer: ```suggestion
Furthermore, you can use the `aura-accent-color` class name to reset the accent color back to global accent color, for example, if a parent container has applied another accent color.
```

*/

.elevated-box {
  /* Border styles are omitted */
	background-color: var(--aura-surface-color);
}
----

image::_images/surface.png[width=300]
====

=== Surface Level (Brightness)

The brightness of the surface color is controlled by the `--aura-surface-level` property – a number that defaults to 1, which creates a brighter or "elevated" surface. Negative numbers create a darker or "recessed" surface. Level 0 is essentially the same color as `--aura-background-color`.

The surface color turns to pure white (with positive levels) or to pure black (with negative levels) after a certain level, depending on `--aura-background-color`.

.Surface colors for levels from -1 to 4, in both light and dark color schemes.
image::_images/surface-colors.png[width=500]

To make the surface look "recessed" (darker), you can set the level, for example, to -1, either in a stylesheet or with an inline style. The number can be any integer or decimal.

.Use the Surface Color Class Names
[IMPORTANT]
You need to use the `aura-surface` or `aura-surface-solid` class name if you want to change the `--aura-surface-level` property. Changing it on other elements doesn't work.

> .Setting the surface level on a generic container element

Reviewer: The code snippets are potentially confusing. It could be improved by:
- Grouping the class name / stylesheet snippets together
- Demonstrating to add the `aura-surface` class name to the inline styles snippet as well

Author: Tried to improve the presentation.

  --aura-accent-color-dark: var(--aura-orange);
	background-color: var(--aura-accent-surface);
}
----

image::_images/accent-surface-orange.png[width=300]
====

The accent surface color can be adjusted with the level and opacity properties the same as the regular surface color.

.Accent surface colors for levels from -1 to 4, in both light and dark color schemes.
image::_images/accent-surface-colors.png[width=500]

=== Surface Opacity

By default, the Aura surface color is semi-transparent, to achieve a brighter or darker surface on top of _any_ background color.

The surface opacity is controlled by the `--aura-surface-opacity` property, which defaults to 0.5 (i.e., 50% opacity). You can set it to any value between 0 (fully transparent) and 1 (fully opaque).

The surface's opaqueness naturally affects its brightening or darkening effect over the background, and the extent to which it blends with the background color.

.Use the Surface Color Class Names
[IMPORTANT]
You need to use the `aura-surface` or `aura-surface-solid` class name if you want to change the `--aura-surface-opacity` property. Changing it on other elements doesn't work.

> .Changing the surface opacity on a container that uses the surface color

Reviewer: Similar issue with the example here

[example]
====
[source,java]
----
var box = new Div();
box.addClassNames("aura-surface", "elevated-box");
----

With inline styles:

[source,java]
----
box.getStyle().set("--aura-surface-opacity", "0.8");
----

With a stylesheet:

[source,css]
----
.elevated-box {
  --aura-surface-opacity: 0.8;
}
----

image::_images/surface-opacity.png[width=300]
====

> A fully opaque surface color can also be applied with the aura-surface-solid classname:

Reviewer: ```suggestion
A fully opaque surface color can also be applied with the `aura-surface-solid` classname:
```

image::_images/content-color-scheme.png[]
====

.Changing the color scheme on all components and elements in the application isn't supported.
[CAUTION]
Applying the `color-scheme` property is only supported on a limited set of components and elements. You can view the link:https://github.com/vaadin/web-components/blob/c8681c44ee581fe51958ca1ae306e8f0d55c6353/packages/aura/src/color.css#L57-L93[list of supported CSS selectors].

=== Ensuring Compatibility with Both Light & Dark Schemes

To ensure that any custom styling you apply to the UI works in the applied color scheme:

> - Use the `-light` and `-dark` suffixed properties, like `--aura-background-color-light` and `-aura-background-color-dark` to _customize_ that color.

Reviewer 2: ```suggestion
- Use the `-light` and `-dark` suffixed properties, like `--aura-background-color-light` and `--aura-background-color-dark` to _customize_ that color.
```

`--aura-accent-text-color-light`::
The accent text color for the light color scheme. Computed from `--aura-accent-color-light` by default. If you use a custom color, make sure it has at least 4.5:1 contrast with the background color.

`--aura-accent-text-color-dark`::
The accent text color for the dark color scheme. Computed from `--aura-accent-color-dark` by default. If you use a custom color, make sure it has at least 4.5:1 contrast with the background color.

`--aura-*-text`::
See <<#saturated-palette,saturated palette colors>>.

[[border-color]]
== Border Color

Aura uses the `--vaadin`-prefixed <<../base#border-color,base border properties>> by assigning its own computed, color-scheme-adapted values to them, but also provides its own <<#accent-color,accent>> border color.

Border colors are by default computed based on other color properties and the <<#contrast,contrast>> level. The most common way to customize border colors is to adjust the contrast level.

[.property-table]
`--aura-accent-border-color` xref:#light-dark-function[light-dark(),role="badge light-dark"]::

> A border color that is slightly tinted with the <<#accent-color>,accent color>, designed to be paired together with the `--aura-accent-surface` color.

Reviewer 2: ```suggestion
A border color that is slightly tinted with the <<#accent-color,accent color>>, designed to be paired together with the `--aura-accent-surface` color.
```

|`--aura-purple` | `--aura-purple-text` xref:#light-dark-function[light-dark(),role="badge light-dark"]
|===

=== Custom Brightness or Opacity Variants

Aura doesn't offer traditional color scales that many design systems include, where you have 10 to 12 colors per hue ranging from almost white to almost black, commonly named from 50/100 to 900, or as `10pct` to `90pct` in the Lumo theme.

The Aura palette consists of single saturated color values, from which all necessary variations of that color are computed automatically (the text, border, and surface/background variants). If you need more than what Aura and the <<../base#,base styles>> offer out of the box, you can use the native CSS https://developer.mozilla.org/en-US/docs/Web/CSS/Reference/Values/color_value/color-mix[`color-mix()`] and https://developer.mozilla.org/en-US/docs/Web/CSS/Guides/Colors/Using_relative_colors[relative color functions] to compute more color properties.

.Example: Computing color variants based on the `--aura-red` property
[example]
====
[source,css]
----
html {
  /* A variant of --aura-red mixed with 20% --aura-background-color */

>   --aura-red-light: color-mix(in srgb, var(--aura-red) 80%, var(--aura-background-color);

Reviewer 2: Missing closing `)`:

```suggestion
  --aura-red-light: color-mix(in srgb, var(--aura-red) 80%, var(--aura-background-color));
```

  /* A variant of --aura-red with 20% opacity */

>   --aura-red-20pct: rgb(var(--aura-red) / 20%);

Reviewer 2: 🤖 says this is wrong, and it seems legit:

> `rgb(var(--aura-red) / 20%)` is invalid CSS. `var(--aura-red)` resolves to a hex/named color, which `rgb()` can't take that way. Either `color-mix(in srgb, var(--aura-red) 20%, transparent)` or relative-color `rgb(from var(--aura-red) r g b / 20%)`.

Author: Yeah, thanks for catching this. I think I copied this as-is from Rolf’s draft content, didn't pay attention.

Author: Decided to use color-mix() to make it easier to understand.

[.property-table]
`--aura-accent-color` xref:#light-dark-function[light-dark(),role="badge light-dark"]::
Automatically assigned the value of the `-light` or `-dark` suffixed property, depending on the active color scheme.

`--aura-accent-color-light`::
The accent color for the light color scheme. The default is `var(--aura-blue)`.

`--aura-accent-color-dark`::
The accent color for the dark color scheme. The default is `var(--aura-blue)`.

`--aura-accent-contrast-color` xref:#light-dark-function[light-dark(),role="badge light-dark"]::
A color that provides strong contrast against the `--aura-accent-color`, for example, button label color when the button background uses the accent color. The default contrast color is either white or a semi-transparent black, depending on the lightness of the accent color.

`--aura-accent-contrast-color-light`::
The accent contrast color for the light color scheme.

`--aura-accent-contrast-color-dark`::
The accent contrast color for the dark color scheme.

.Define accent colors for the light and dark color schemes
[example]
====
[source,css]
----

> /* You can apply different colors for light and dark color schemes */

Reviewer 2: Custom CSS properties below need to be changed to replace incorrectly used double dash: `--aura--accent` - should be `--aura-accent-`.

Author: Done

html {
  --aura-accent-color-light: var(--aura-orange);
  --aura-accent-color-dark: var(--aura-purple);
}
----

.Using the orange accent color in the light color scheme and the purple accent color in the dark color scheme
image::_images/accent-colors.png[]
====

=== Accent Color Class Names

The accent color is applied globally to the entire UI, but you can also scope it to individual elements.

Aura provides the following CSS utility classes, based on the <<#saturated-palette,saturated color palette>>, that you can apply to any part of the UI to give it a different accent color:

- `aura-accent-neutral`
- `aura-accent-red`
- `aura-accent-orange`
- `aura-accent-yellow`
- `aura-accent-green`
- `aura-accent-blue`
- `aura-accent-purple`

// .Use the purple palette color as the global accent color
// [example]
// ====
// [source,html]
// ----
// <html class="aura-accent-purple">
// ----

> // ====

Reviewer 2: Is this commented code needed or should it be removed?

Author: Removed.

.Setting the surface level on a component
[example]
====
[source,java]
----
var box = new Div();
box.addClassNames("aura-surface", "recessed-box");
----

You can choose to set the surface level with either inline styles in Java, or with CSS in a stylesheet:

[.example]
--
[source,java]
----
<source-info group="Java"></source-info>
box.getStyle().set("--aura-surface-level", "-1");
// Border styles are omitted
----

.styles.css
[source,css]
----
<source-info group="CSS"></source-info>
.recessed-box {
  --aura-surface-level: -1;
  /* Border styles are omitted */
}
----
--

image::_images/surface-level.png[width=300]
====

.Surface level is inherited by child elements.
[NOTE]
Any child element to which the Aura surface color is applied inherits the surface level of its parent. You can override that by applying a different surface level on the child element.

=== Accent Surface

A more colorful variant of the surface color can be achieved with the `--aura-accent-surface` property, which blends the surface color with the <<#accent-color,accent color>>.

// .Only saturated colors affect the accent surface.
// [NOTE]

> // The effective accent color needs to have saturation (i.e., not white/gray/black) for it to produce a different color than the regular surface color.

Reviewer 2: Is this commented code needed or should it be removed?

Author: I wasn't sure if this would be useful or not. Let’s see. Removed for now.

An opaque version of the effective surface color. Essentially the same as if `--aura-surface-opacity` was `1`. Useful for situations when a semi-transparent color is not suitable (e.g., when a surface needs to fully cover something).

`--aura-accent-surface` xref:./#read-only-properties[Read-only,role=badge]::
A surface color that is tinted with the effective <<#accent-color,accent color>>.

`--aura-surface-level`::
The “elevation level” of the computed surface color. Can be any fractional number, positive, or negative. The default is `1`. The level is inherited by nested surfaces.
+
With the light color scheme, levels 3–4 usually result in a white color, depending on the lightness of `--aura-background-color-light`.
+
With the dark color scheme, level 8 starts to be an upper limit, after which text colors are likely to not have enough <<#contrast,contrast>> (depending on the surface opacity).

`--aura-surface-opacity`::
The transparency of the computed surface color. The default is `0.5`, and each surface resets its opacity to this value unless explicitly overridden. Transparency allows you to nest the same surface color on top of each other to create more sense of elevation.

> // [NOTE]

Reviewer 2: This whole part is commented - should it also be removed?

Author: Yeah, removed. I was keeping it there only to potentially reuse some of it later, if needed.
